### ISSUE TITLE
Where is the corresponding code for equation [7-10]

### ISSUE THREAD

**Issue by hhhusiyi-monash** [Hwhitetooth/lirpg#2]
Hi

Thanks for your awesome work,

In your code, I only found the code for equation 11 which is the first term of equation 5.
However, I can not find the corresponding part for equation [7-10] which is the second term of equation 5.
Could  you show me pls?

Best.
C

**Comment by Hwhitetooth** [comment]
Hi @hhhusiyi-monash,

Equations 7-10 are presented for a better explanation of the derivation. In practice, we can use modern audodiff tools like TensorFlow to compute the meta-gradient in one backward pass. The computation denoted by Equation 7-10 are implicitly done by TensorFlow. In this code, we directly computed Equation 5, which is exactly what we need.

Best regards,
Zeyu

**Comment by boscotsang** [comment]
> 
> 
> Hi @hhhusiyi-monash,
> 
> Equations 7-10 are presented for a better explanation of the derivation. In practice, we can use modern audodiff tools like TensorFlow to compute the meta-gradient in one backward pass. The computation denoted by Equation 7-10 are implicitly done by TensorFlow. In this code, we directly computed Equation 5, which is exactly what we need.
> 
> Best regards,
> Zeyu

```
        neglogpac_new = policy_new.pd.neglogp(A)
        ratio_new = tf.exp(OLDNEGLOGPAC - neglogpac_new)
        pg_ex_loss1 = -ADV_EX * ratio_new
        pg_ex_loss2 = -ADV_EX * tf.clip_by_value(ratio_new, 1.0 - CLIPRANGE, 1.0 + CLIPRANGE)
        pg_ex_loss = tf.reduce_mean(tf.maximum(pg_ex_loss1, pg_ex_loss2))
        v_ex = train_model.v_ex
        v_ex_clipped = OLDV_EX + tf.clip_by_value(v_ex - OLDV_EX, - CLIPRANGE, CLIPRANGE)
        v_ex_loss1 = tf.square(v_ex - RET_EX)
        v_ex_loss2 = tf.square(v_ex_clipped - RET_EX)
        v_ex_loss = .5 * tf.reduce_mean(tf.maximum(v_ex_loss1, v_ex_loss2))
        intrinsic_loss = pg_ex_loss + vf_coef * v_ex_loss
        intrinsic_params = tf.trainable_variables("intrinsic")
        intrinsic_grads = tf.gradients(intrinsic_loss, intrinsic_params)
        if max_grad_norm is not None:
            intrinsic_grads, intrinsic_grad_norm = tf.clip_by_global_norm(intrinsic_grads, max_grad_norm)
        intrinsic_grads_and_vars = list(zip(intrinsic_grads, intrinsic_params))
        intrinsic_trainer = tf.train.AdamOptimizer(learning_rate=LR_BETA, epsilon=1E-5)
        intrinsic_train = intrinsic_trainer.apply_gradients(intrinsic_grads_and_vars)
```

It seems the the code above calculate the gradient w.r.t intrinsic_params. Instrinsic_loss is not affected by intrinsic_params and it only depends on extrinsic policy loss and extrinsic value loss.

**Comment by Hwhitetooth** [comment]
Note that the extrinsic policy gradient loss is computed with the _updated_ policy parameters. The inner loop uses intrinsic rewards to update the policy. Thus there is an indirect dependency between the extrinsic policy gradient loss and the intrinsic reward parameters.

In terms of implementation, in order to make sure the gradient flows back from the extrinsic policy gradient loss to the intrinsic reward parameters _through_ the updated policy, the updated policy is constructed by `Tensors` instead of `Variables`. The corresponding code can be found at
https://github.com/Hwhitetooth/lirpg/blob/master/baselines/ppo2/ppo2.py#L71
and
https://github.com/Hwhitetooth/lirpg/blob/master/baselines/ppo2/policies.py#L235

**Comment by yalidu** [comment]
The gradients to 'intrinsic reward' parameters are always zero because of the feed dict o.
https://github.com/Hwhitetooth/lirpg/blob/ca85c523beae4feb0fa128c90f145961a9d0de36/baselines/ppo2/policies.py#L188

**Comment by AlexZhaoZt** [comment]
Hi,

I am confused by the "manual" update of policy (under https://github.com/Hwhitetooth/lirpg/blob/master/baselines/ppo2/ppo2.py#L65). It looks like the policy parameters are updated twice because `apply_gradients` is called above it.
 
I was able to come up with a possible reason; can you help me verify if this is correct? The policy parameters are not updated when `apply_gradients` is called, but rather when `sess.run` is called. Thus, the "manual" update of policy does not apply the gradients twice.

**Comment by Hwhitetooth** [comment]
Hi @AlexZhaoZt 

Yes, your explanation is correct. In TensorFlow, `apply_gradients` only creates a node in the computation graph but does not really execute it. The actual update only happens when `sess.run` is called. The reason I did `apply_gradients` before constructing the updated policy was to get the optimizer states.

**Comment by boscotsang** [comment]
> Note that the extrinsic policy gradient loss is computed with the _updated_ policy parameters. The inner loop uses intrinsic rewards to update the policy. Thus there is an indirect dependency between the extrinsic policy gradient loss and the intrinsic reward parameters.
> 
> In terms of implementation, in order to make sure the gradient flows back from the extrinsic policy gradient loss to the intrinsic reward parameters _through_ the updated policy, the updated policy is constructed by `Tensors` instead of `Variables`. The corresponding code can be found at
> https://github.com/Hwhitetooth/lirpg/blob/master/baselines/ppo2/ppo2.py#L71
> and
> https://github.com/Hwhitetooth/lirpg/blob/master/baselines/ppo2/policies.py#L235

So the policy prameter and intrinsic reward parameter didn't have direct gradient flow. We update the policy first and then use the new policy to calculate the gradients of instrinsic reward parameter.
